src/libs/sfdk/dockervirtualmachine.cpp

```diff
@@ -0,0 +1,529 @@
+}
+
+void DockerVirtualMachinePrivate::doAddPortForwarding(const QString &ruleName,
+        const QString &protocol, quint16 hostPort, quint16 emulatorVmPort, const QObject *context,
+        const Functor<bool> &functor)
+{
+    Q_UNUSED(ruleName)
+    Q_UNUSED(protocol)
+    Q_UNUSED(hostPort)
+    Q_UNUSED(emulatorVmPort)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doRemovePortForwarding(const QString &ruleName,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_UNUSED(ruleName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortForwarding(
+        VirtualMachinePrivate::ReservedPort which, quint16 port, const QObject *context,
+        const Functor<bool> &functor)
+{
+    buildWithLabel(labelFor(which), QString::number(port), context, functor);
+}
+
+void DockerVirtualMachinePrivate::doSetReservedPortListForwarding(
+        VirtualMachinePrivate::ReservedPortList which, const QList<Utils::Port> &ports,
+        const QObject *context, const Functor<const QMap<QString, quint16> &, bool> &functor)
+{
+    Q_UNUSED(which)
+    Q_UNUSED(ports)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+void DockerVirtualMachinePrivate::doRestoreSnapshot(const QString &snapshotName,
+        const QObject *context, const Functor<bool> &functor)
+{
+    Q_UNUSED(snapshotName)
+    Q_UNUSED(context)
+    Q_UNUSED(functor)
+    QTC_CHECK(false);
+}
+
+VirtualMachineInfo DockerVirtualMachinePrivate::virtualMachineInfoFromOutput(
+        const QByteArray &output) const
+{
+    QJsonDocument document = QJsonDocument::fromJson(output);
+    VirtualMachineInfo info = {};
+
+    QTC_ASSERT(!document.isNull(), return info);
+
+    QJsonObject labels = document.object();
+
+    auto parsePort = [=](VirtualMachinePrivate::ReservedPort which, quint16 *port) {
+        QString key = labelFor(which);
+        if (labels.contains(key))
+            *port = labels.value(key).toString().toUShort();
+    };
+
+    parsePort(VirtualMachinePrivate::SshPort, &info.sshPort);
+    parsePort(VirtualMachinePrivate::WwwPort, &info.wwwPort);
+
+    auto parsePath = [=](VirtualMachinePrivate::SharedPath which, QString *path) {
+        QString key = labelFor(which);
+        if (labels.contains(key))
+            *path = labels.value(key).toString();
+    };
+
+    parsePath(VirtualMachinePrivate::SharedSrc, &info.sharedSrc);
+    parsePath(VirtualMachinePrivate::SharedSsh, &info.sharedSsh);
+    parsePath(VirtualMachinePrivate::SharedHome, &info.sharedHome);
+    parsePath(VirtualMachinePrivate::SharedConfig, &info.sharedConfig);
+    parsePath(VirtualMachinePrivate::SharedTargets, &info.sharedTargets);
+
+    info.memorySizeMb = VirtualMachine::availableMemorySizeMb();
+    info.cpuCount = VirtualMachine::availableCpuCount();
```

Comment:
I would consider moving these two near the beginning

Reply:
fixed